insights/parsers/nginx_conf.py

```diff
@@ -0,0 +1,335 @@
+
+    mail {
+      server_name mail.example.com;
+      auth_http  localhost:9000/cgi-bin/auth;
+      server {
+        listen   143;
+        protocol imap;
+      }
+    }
+
+    http {
+      include  /etc/nginx/conf.d/*.conf
+      index    index.html index.htm index.php;
+
+      default_type application/octet-stream;
+      log_format   main '$remote_addr - $remote_user [$time_local]  $status '
+                        '"$request" $body_bytes_sent "$http_referer" '
+                        '"$http_user_agent" "$http_x_forwarded_for"';
+      access_log   logs/access.log  main;
+      sendfile     on;
+      tcp_nopush   on;
+      server_names_hash_bucket_size 128;
+
+      server { # php/fastcgi
+        listen       80;
+        server_name  domain1.com www.domain1.com;
+        access_log   logs/domain1.access.log  main;
+        root         html;
+
+        location ~ \.php$ {
+          fastcgi_pass   127.0.0.1:1025;
+        }
+      }
+
+      server { # simple reverse-proxy
+        listen       80;
+        server_name  domain2.com www.domain2.com;
+        access_log   logs/domain2.access.log  main;
+
+        location ~ ^/(images|javascript|js|css|flash|media|static)/  {
+          root    /var/www/virtual/big.server.com/htdocs;
+          expires 30d;
+        }
+
+        location / {
+          proxy_pass   http://127.0.0.1:8080;
+        }
+      }
+
+      map $http_upgrade $connection_upgrade {
+        default upgrade;
+        '' close;
+      }
+
+      upstream websocket {
+        server 10.66.208.205:8010;
+      }
+
+      upstream big_server_com {
+        server 127.0.0.3:8000 weight=5;
+        server 127.0.0.3:8001 weight=5;
+        server 192.168.0.1:8000;
+        server 192.168.0.1:8001;
+      }
+
+      server { # simple load balancing
+        listen          80;
+        server_name     big.server.com;
+        access_log      logs/big.server.access.log main;
+
+        location / {
+          proxy_pass      http://big_server_com;
+        }
+      }
+    }
+
+Examples:
+
+    >>> nginxconf = NginxConf(context_wrap(NGINXCONF))
+    >>> nginxconf['user']
+    'root'
+    >>> nginxconf['events']['worker_connections'] # Values are all kept as strings.
+    '4096'
+    >>> nginxconf['mail']['server']['listen']
+    '143'
+    >>> nginxconf['http']['access_log']
+    'logs/access.log  main'
+    >>> nginxconf['http']['server'][0]['location']['fastcgi_pass']
+    '127.0.0.1:1025'
+"""
+import string
+import copy
+from .. import parser, LegacyItemAccess, Parser, get_active_lines
+from insights.contrib.pyparsing import (
+    Literal, White, Word, alphanums, CharsNotIn, Combine, Forward, Group,
+    Optional, OneOrMore, ZeroOrMore, Regex, stringEnd, restOfLine)
+
+
+def create_parser():
+    """
+    A class that parses nginx configuration with pyparsing
+
+    The MIT License (MIT)
+
+    Copyright (c) 2014 Fatih Erikli
+
+    Permission is hereby granted, free of charge, to any person obtaining a copy of
```

Comment:
Just a note here - you can't apply the MIT license to just a part of a file AFAICS.  It applies to the whole file, and possibly the whole project.

If you're using code that you can't import from a standard location - i.e. that isn't installed from RPM package or via pip - then the usual form AFAIK is to say something like:
```
This parser is derived from the code by Fatih Erikli at:

http://url.to/code/

Used under the MIT license.
```
Check with someone else though - this may just be what I'm used to.

This also allows us to check on the upstream code for updates and possibly supply bug fixes.  We should always try to work directly with upstream where possible.

Reply:
@bfahr - Do we have a standard workflow to handle open source code reference? Add license  to the function docstring, or do as Paul said?

Reply:
@klape @jhjaggars how do you recommend we handle adding code from other open source projects?

Reply:
@wushiqinlou @jhjaggars recommends that this function be placed in the `insights/contrib` directory.  Please backport this change to the 1.x branch as well.  Take a look at the modules that are already there for examples.